src/Redis/Orleans.Persistence.Redis/Storage/RedisGrainStorage.cs

```diff
@@ -195,19 +202,18 @@ public async Task WriteStateAsync<T>(string grainType, GrainId grainId, IGrainSt
 
             if (writeWithScriptResponse is not null && writeWithScriptResponse.IsNull)
             {
-                throw new InconsistentStateException($"ETag mismatch - tried with ETag: {grainState.ETag}");
+                throw new InconsistentStateException(Invariant($"ETag mismatch - tried with ETag: {grainState.ETag}"));
             }
 
             grainState.ETag = newEtag;
         }
 
-        private Task<RedisResult> WriteToRedisUsingPreparedScriptAsync(RedisValue payload, string etag, string key, string newEtag)
+        private Task<RedisResult> WriteToRedisUsingPreparedScriptAsync(RedisValue payload, string etag, RedisKey key, string newEtag)
         {
             var keys = new RedisKey[] { key };
-            var args = new RedisValue[] { etag, newEtag, payload };
+            var args = new RedisValue[] { etag, newEtag, payload, _ttl };
```

Comment:
note to self: we should add an overload that takes `ReadOnlyMemory<RedisKey> keys, ReadOnlyMemory<RedisValue> values` so that we could allow pooled use here; also, I wonder how many times a script takes only one key? I'd wager "very high %" - maybe there's a valid justification for `RedisKey key, ReadOnlyMemory<RedisValue> values = default` (thoughts @NickCraver @philon-msft  ?)

Reply:
I just woke up, but how would `ReadOnlyMemory<T>` help us? It seems we'd still need to allocate some backing storage (an array). We'd need `ReadOnlySpan<T>`, which has its issues with async/await (which can be worked around at some cost to readability)

Reply:
``` c#
var args= ArrayPool<RedisValue>.Shared.Rent(4); // note: oversized
args[0] = etag;
args[1] = newEtag;
args[2] = payload;
args[3] = _ttl;
//
... blah ... new ReadOnlyMemory<RedisVaue>(args, 0, 4) ...
//
ArrayPool<RedisValue>.Shared.Return(args);
```

Reply:
for consideration: https://github.com/StackExchange/StackExchange.Redis/pull/2346

Reply:
When the changes are in upstream, we will adopt them